Flaky unit test TestRespawn in supervisor package

> supervisor_test.go:155: Respawn failed: supervisor-test-respawn

I've noticed that a few times now, both on my box, as well as on CI.

<details>
<summary>Build log snippets from a recent failure</summary>

https://github.com/k0sproject/k0s/runs/6348983779?check_suite_focus=true

```text
2022-05-09T09:16:31.9039564Z ##[group]Run make check-unit
2022-05-09T09:16:31.9040430Z [36;1mmake check-unit[0m
2022-05-09T09:16:31.9065877Z shell: /usr/bin/bash -e {0}
2022-05-09T09:16:31.9066597Z env:
2022-05-09T09:16:31.9067375Z   GO_VERSION: 1.18.1
2022-05-09T09:16:31.9068085Z   GO_TEST_RACE: 
2022-05-09T09:16:31.9068746Z ##[endgroup]
2022-05-09T09:16:32.7389005Z GOCACHE='/home/gh-runner/actions-runner/_work/k0s/k0s/build/cache/go/build' GOMODCACHE='/home/gh-runner/actions-runner/_work/k0s/k0s/build/cache/go/mod' go mod tidy
2022-05-09T09:16:33.7889518Z GOCACHE='/home/gh-runner/actions-runner/_work/k0s/k0s/build/cache/go/build' GOMODCACHE='/home/gh-runner/actions-runner/_work/k0s/k0s/build/cache/go/mod' go test  `GOCACHE='/home/gh-runner/actions-runner/_work/k0s/k0s/build/cache/go/build' GOMODCACHE='/home/gh-runner/actions-runner/_work/k0s/k0s/build/cache/go/mod' go list . ./cmd/... ./pkg/... ./internal/... ./static/... ./hack/...`
2022-05-09T09:19:20.0686970Z ?   	github.com/k0sproject/k0s	[no test files]
2022-05-09T09:19:20.0688214Z ?   	github.com/k0sproject/k0s/cmd	[no test files]
2022-05-09T09:19:24.4404375Z ok  	github.com/k0sproject/k0s/cmd/airgap	0.076s
2022-05-09T09:19:24.4407621Z ?   	github.com/k0sproject/k0s/cmd/api	[no test files]
2022-05-09T09:19:24.4410483Z ?   	github.com/k0sproject/k0s/cmd/backup	[no test files]
2022-05-09T09:19:24.4413322Z ?   	github.com/k0sproject/k0s/cmd/config	[no test files]
2022-05-09T09:19:24.4416968Z ?   	github.com/k0sproject/k0s/cmd/controller	[no test files]
2022-05-09T09:19:24.4419944Z ?   	github.com/k0sproject/k0s/cmd/ctr	[no test files]
2022-05-09T09:19:24.4422793Z ?   	github.com/k0sproject/k0s/cmd/etcd	[no test files]
2022-05-09T09:19:24.4425723Z ?   	github.com/k0sproject/k0s/cmd/install	[no test files]
2022-05-09T09:19:36.0757145Z ok  	github.com/k0sproject/k0s/cmd/kubeconfig	3.072s
2022-05-09T09:19:36.0758417Z ?   	github.com/k0sproject/k0s/cmd/kubectl	[no test files]
2022-05-09T09:19:36.0759599Z ?   	github.com/k0sproject/k0s/cmd/reset	[no test files]
2022-05-09T09:19:36.0760687Z ?   	github.com/k0sproject/k0s/cmd/restore	[no test files]
2022-05-09T09:19:36.0765337Z ?   	github.com/k0sproject/k0s/cmd/start	[no test files]
2022-05-09T09:19:36.0766491Z ?   	github.com/k0sproject/k0s/cmd/status	[no test files]
2022-05-09T09:19:36.0767822Z ?   	github.com/k0sproject/k0s/cmd/stop	[no test files]
2022-05-09T09:19:36.0768975Z ok  	github.com/k0sproject/k0s/cmd/sysinfo	0.039s
2022-05-09T09:19:36.0770083Z ?   	github.com/k0sproject/k0s/cmd/token	[no test files]
2022-05-09T09:19:36.0771234Z ?   	github.com/k0sproject/k0s/cmd/validate	[no test files]
2022-05-09T09:19:36.0772378Z ?   	github.com/k0sproject/k0s/cmd/version	[no test files]
2022-05-09T09:19:36.0773498Z ?   	github.com/k0sproject/k0s/cmd/worker	[no test files]
2022-05-09T09:19:36.0788802Z ?   	github.com/k0sproject/k0s/pkg/airgap	[no test files]
2022-05-09T09:19:36.0797434Z ?   	github.com/k0sproject/k0s/pkg/apis/helm.k0sproject.io/v1beta1	[no test files]
2022-05-09T09:19:36.0798846Z ?   	github.com/k0sproject/k0s/pkg/apis/k0s.k0sproject.io/clientset	[no test files]
2022-05-09T09:19:36.0800170Z ?   	github.com/k0sproject/k0s/pkg/apis/k0s.k0sproject.io/clientset/fake	[no test files]
2022-05-09T09:19:36.0801811Z ?   	github.com/k0sproject/k0s/pkg/apis/k0s.k0sproject.io/clientset/scheme	[no test files]
2022-05-09T09:19:36.0803301Z ?   	github.com/k0sproject/k0s/pkg/apis/k0s.k0sproject.io/clientset/typed/k0s.k0sproject.io/v1beta1	[no test files]
2022-05-09T09:19:36.0804921Z ?   	github.com/k0sproject/k0s/pkg/apis/k0s.k0sproject.io/clientset/typed/k0s.k0sproject.io/v1beta1/fake	[no test files]
2022-05-09T09:19:36.0806326Z ok  	github.com/k0sproject/k0s/pkg/apis/k0s.k0sproject.io/v1beta1	0.188s
2022-05-09T09:19:38.5483910Z ok  	github.com/k0sproject/k0s/pkg/applier	0.187s
2022-05-09T09:19:38.5485026Z ?   	github.com/k0sproject/k0s/pkg/assets	[no test files]
2022-05-09T09:19:38.5486519Z ?   	github.com/k0sproject/k0s/pkg/backup	[no test files]
2022-05-09T09:19:38.5492036Z ?   	github.com/k0sproject/k0s/pkg/build	[no test files]
2022-05-09T09:19:38.5493209Z ?   	github.com/k0sproject/k0s/pkg/certificate	[no test files]
2022-05-09T09:19:38.5494282Z ?   	github.com/k0sproject/k0s/pkg/cleanup	[no test files]
2022-05-09T09:19:38.5495334Z ok  	github.com/k0sproject/k0s/pkg/component	0.049s
2022-05-09T09:19:41.5958128Z ok  	github.com/k0sproject/k0s/pkg/component/controller	2.110s
2022-05-09T09:19:41.5959700Z ?   	github.com/k0sproject/k0s/pkg/component/controller/clusterconfig	[no test files]
2022-05-09T09:19:41.5961002Z ?   	github.com/k0sproject/k0s/pkg/component/status	[no test files]
2022-05-09T09:19:41.5962147Z ok  	github.com/k0sproject/k0s/pkg/component/worker	0.175s
2022-05-09T09:19:41.5963252Z ok  	github.com/k0sproject/k0s/pkg/config	0.223s
2022-05-09T09:19:41.5964296Z ok  	github.com/k0sproject/k0s/pkg/constant	0.036s
2022-05-09T09:19:41.5965475Z ?   	github.com/k0sproject/k0s/pkg/container/runtime	[no test files]
2022-05-09T09:19:41.5966721Z ok  	github.com/k0sproject/k0s/pkg/debounce	2.040s
2022-05-09T09:19:41.5967995Z ?   	github.com/k0sproject/k0s/pkg/etcd	[no test files]
2022-05-09T09:19:41.5969030Z ?   	github.com/k0sproject/k0s/pkg/helm	[no test files]
2022-05-09T09:19:41.5970075Z ?   	github.com/k0sproject/k0s/pkg/install	[no test files]
2022-05-09T09:19:41.5971142Z ok  	github.com/k0sproject/k0s/pkg/k0scloudprovider	0.170s
2022-05-09T09:19:41.5972215Z ok  	github.com/k0sproject/k0s/pkg/kubernetes	0.057s
2022-05-09T09:19:41.5973274Z ok  	github.com/k0sproject/k0s/pkg/leaderelection	7.735s
2022-05-09T09:19:41.5974368Z ?   	github.com/k0sproject/k0s/pkg/performance	[no test files]
2022-05-09T09:19:41.5976951Z time="2022-05-09T05:19:30-04:00" level=info msg="Starting to supervise" component=supervisor-test-sleep
2022-05-09T09:19:41.5978716Z time="2022-05-09T05:19:30-04:00" level=info msg="Started successfully, go nuts" component=supervisor-test-sleep
2022-05-09T09:19:41.5980439Z time="2022-05-09T05:19:30-04:00" level=info msg="Shutting down pid 12890" component=supervisor-test-sleep
2022-05-09T09:19:41.5982146Z time="2022-05-09T05:19:30-04:00" level=info msg="Starting to supervise" component=supervisor-test-fail
2022-05-09T09:19:41.5983840Z time="2022-05-09T05:19:30-04:00" level=info msg="Started successfully, go nuts" component=supervisor-test-fail
2022-05-09T09:19:41.5985825Z time="2022-05-09T05:19:30-04:00" level=info msg="Shutting down pid 12892" component=supervisor-test-fail
2022-05-09T09:19:41.5987636Z time="2022-05-09T05:19:30-04:00" level=info msg="Starting to supervise" component=supervisor-test-non-executable
2022-05-09T09:19:41.5989666Z time="2022-05-09T05:19:30-04:00" level=warning msg="Failed to start: fork/exec /tmp: permission denied" component=supervisor-test-non-executable
2022-05-09T09:19:41.5991790Z time="2022-05-09T05:19:30-04:00" level=warning msg="failed to initialize dir: mkdir /bin/sh: not a directory" component=supervisor-test-rundir-fail
2022-05-09T09:19:41.5993693Z time="2022-05-09T05:19:30-04:00" level=info msg="Starting to supervise" component=supervisor-test-respawn
2022-05-09T09:19:41.5995464Z time="2022-05-09T05:19:30-04:00" level=info msg="Started successfully, go nuts" component=supervisor-test-respawn
2022-05-09T09:19:41.5997256Z time="2022-05-09T05:19:30-04:00" level=warning msg="Process exited with code: 0" component=supervisor-test-respawn
2022-05-09T09:19:41.5998948Z time="2022-05-09T05:19:30-04:00" level=info msg="respawning in 10ms" component=supervisor-test-respawn
2022-05-09T09:19:41.6000553Z time="2022-05-09T05:19:30-04:00" level=info msg="Restarted (1)" component=supervisor-test-respawn
2022-05-09T09:19:41.6002217Z time="2022-05-09T05:19:30-04:00" level=info msg="Shutting down pid 12895" component=supervisor-test-respawn
2022-05-09T09:19:41.6003470Z --- FAIL: TestRespawn (0.01s)
2022-05-09T09:19:41.6004682Z     supervisor_test.go:155: Respawn failed: supervisor-test-respawn
2022-05-09T09:19:41.6006463Z time="2022-05-09T05:19:30-04:00" level=info msg="Starting to supervise" component=supervisor-test-stop-while-respawn
2022-05-09T09:19:41.6009436Z time="2022-05-09T05:19:30-04:00" level=info msg="Started successfully, go nuts" component=supervisor-test-stop-while-respawn
2022-05-09T09:19:41.6011427Z time="2022-05-09T05:19:30-04:00" level=warning msg="exit status 1" component=supervisor-test-stop-while-respawn
2022-05-09T09:19:41.6013278Z time="2022-05-09T05:19:30-04:00" level=info msg="respawning in 1s" component=supervisor-test-stop-while-respawn
2022-05-09T09:19:41.6014362Z FAIL
2022-05-09T09:19:41.6015197Z FAIL	github.com/k0sproject/k0s/pkg/supervisor	0.070s
2022-05-09T09:19:41.6016260Z ?   	github.com/k0sproject/k0s/pkg/telemetry	[no test files]
2022-05-09T09:19:41.6017304Z ?   	github.com/k0sproject/k0s/pkg/token	[no test files]
2022-05-09T09:19:41.6018392Z ?   	github.com/k0sproject/k0s/internal/pkg/archive	[no test files]
2022-05-09T09:19:41.6019474Z ok  	github.com/k0sproject/k0s/internal/pkg/dir	0.032s
2022-05-09T09:19:41.6020505Z ok  	github.com/k0sproject/k0s/internal/pkg/file	0.013s
2022-05-09T09:19:41.6021591Z ok  	github.com/k0sproject/k0s/internal/pkg/flags	0.012s
2022-05-09T09:19:41.6022694Z ?   	github.com/k0sproject/k0s/internal/pkg/iface	[no test files]
2022-05-09T09:19:41.6023826Z ?   	github.com/k0sproject/k0s/internal/pkg/random	[no test files]
2022-05-09T09:19:41.6024967Z ok  	github.com/k0sproject/k0s/internal/pkg/strictyaml	0.028s
2022-05-09T09:19:41.6026420Z ok  	github.com/k0sproject/k0s/internal/pkg/stringmap	0.038s
2022-05-09T09:19:41.6027528Z ok  	github.com/k0sproject/k0s/internal/pkg/stringslice	0.031s
2022-05-09T09:19:41.6028646Z ?   	github.com/k0sproject/k0s/internal/pkg/sysinfo	[no test files]
2022-05-09T09:19:41.6029807Z ok  	github.com/k0sproject/k0s/internal/pkg/sysinfo/machineid	0.043s
2022-05-09T09:19:41.6030968Z ok  	github.com/k0sproject/k0s/internal/pkg/sysinfo/probes	0.043s
2022-05-09T09:19:41.6032148Z ok  	github.com/k0sproject/k0s/internal/pkg/sysinfo/probes/linux	0.104s
2022-05-09T09:19:41.6033359Z ?   	github.com/k0sproject/k0s/internal/pkg/templatewriter	[no test files]
2022-05-09T09:19:41.6034527Z ok  	github.com/k0sproject/k0s/internal/pkg/users	0.013s
2022-05-09T09:19:41.6035603Z ?   	github.com/k0sproject/k0s/internal/testutil	[no test files]
2022-05-09T09:19:41.6036749Z ?   	github.com/k0sproject/k0s/internal/testutil/sysinfo	[no test files]
2022-05-09T09:19:41.6037808Z ?   	github.com/k0sproject/k0s/static	[no test files]
2022-05-09T09:19:41.6039159Z ?   	github.com/k0sproject/k0s/hack/validate-images	[no test files]
2022-05-09T09:19:41.6040051Z FAIL
2022-05-09T09:19:42.6002613Z make: *** [Makefile:178: check-unit] Error 1
2022-05-09T09:19:42.6042396Z ##[error]Process completed with exit code 2.
```
</details>

Seen this error again during [the build](https://github.com/k0sproject/k0s/runs/6451837048?check_suite_focus=true) of cc06d3736bafd4a599e70b67e49ed4d7027aab0a. Seems to happen predominantly/only on arm64.

<details>
<summary>Build log snippet</summary>

```text
time="2022-05-16T08:46:56-04:00" level=info msg="Starting to supervise" component=supervisor-test-sleep
time="2022-05-16T08:46:56-04:00" level=info msg="Started successfully, go nuts" component=supervisor-test-sleep
time="2022-05-16T08:46:56-04:00" level=info msg="Shutting down pid 9513" component=supervisor-test-sleep
time="2022-05-16T08:46:57-04:00" level=info msg="Starting to supervise" component=supervisor-test-fail
time="2022-05-16T08:46:57-04:00" level=info msg="Started successfully, go nuts" component=supervisor-test-fail
time="2022-05-16T08:46:57-04:00" level=info msg="Shutting down pid 9569" component=supervisor-test-fail
time="2022-05-16T08:46:57-04:00" level=info msg="Starting to supervise" component=supervisor-test-non-executable
time="2022-05-16T08:46:57-04:00" level=warning msg="Failed to start: fork/exec /tmp: permission denied" component=supervisor-test-non-executable
time="2022-05-16T08:46:57-04:00" level=warning msg="failed to initialize dir: mkdir /bin/sh: not a directory" component=supervisor-test-rundir-fail
time="2022-05-16T08:46:57-04:00" level=info msg="Starting to supervise" component=supervisor-test-respawn
time="2022-05-16T08:46:57-04:00" level=info msg="Started successfully, go nuts" component=supervisor-test-respawn
time="2022-05-16T08:46:57-04:00" level=warning msg="Process exited with code: 0" component=supervisor-test-respawn
time="2022-05-16T08:46:57-04:00" level=info msg="respawning in 10ms" component=supervisor-test-respawn
time="2022-05-16T08:46:57-04:00" level=info msg="Restarted (1)" component=supervisor-test-respawn
time="2022-05-16T08:46:57-04:00" level=info msg="Shutting down pid 9573" component=supervisor-test-respawn
--- FAIL: TestRespawn (0.03s)
    supervisor_test.go:155: Respawn failed: supervisor-test-respawn
time="2022-05-16T08:46:57-04:00" level=info msg="Starting to supervise" component=supervisor-test-stop-while-respawn
time="2022-05-16T08:46:57-04:00" level=info msg="Started successfully, go nuts" component=supervisor-test-stop-while-respawn
time="2022-05-16T08:46:57-04:00" level=warning msg="exit status 1" component=supervisor-test-stop-while-respawn
time="2022-05-16T08:46:57-04:00" level=info msg="respawning in 1s" component=supervisor-test-stop-while-respawn
FAIL
FAIL	github.com/k0sproject/k0s/pkg/supervisor	1.119s
[...]
FAIL
make: *** [Makefile:178: check-unit] Error 1
Error: Process completed with exit code 2.
```
</details>

Seen this error again during [the build](https://github.com/k0sproject/k0s/runs/6659855696?check_suite_focus=true) of 1b47e395bdf5d1db2f130664dd0c3a31bd0ba9e3 on arm.

<details>
<summary>Build log snippet</summary>

```text
Mon, 30 May 2022 18:52:38 GMT time="2022-05-30T14:52:26-04:00" level=info msg="Starting to supervise" component=supervisor-test-sleep
Mon, 30 May 2022 18:52:38 GMT time="2022-05-30T14:52:26-04:00" level=info msg="Started successfully, go nuts" component=supervisor-test-sleep
Mon, 30 May 2022 18:52:38 GMT time="2022-05-30T14:52:26-04:00" level=info msg="Shutting down pid 9627" component=supervisor-test-sleep
Mon, 30 May 2022 18:52:38 GMT time="2022-05-30T14:52:28-04:00" level=info msg="Starting to supervise" component=supervisor-test-fail
Mon, 30 May 2022 18:52:38 GMT time="2022-05-30T14:52:28-04:00" level=info msg="Started successfully, go nuts" component=supervisor-test-fail
Mon, 30 May 2022 18:52:38 GMT time="2022-05-30T14:52:28-04:00" level=info msg="Shutting down pid 9776" component=supervisor-test-fail
Mon, 30 May 2022 18:52:38 GMT time="2022-05-30T14:52:28-04:00" level=info msg="Starting to supervise" component=supervisor-test-non-executable
Mon, 30 May 2022 18:52:38 GMT time="2022-05-30T14:52:28-04:00" level=warning msg="Failed to start: fork/exec /tmp: permission denied" component=supervisor-test-non-executable
Mon, 30 May 2022 18:52:38 GMT time="2022-05-30T14:52:28-04:00" level=warning msg="failed to initialize dir: mkdir /bin/sh: not a directory" component=supervisor-test-rundir-fail
Mon, 30 May 2022 18:52:38 GMT time="2022-05-30T14:52:28-04:00" level=info msg="Starting to supervise" component=supervisor-test-respawn
Mon, 30 May 2022 18:52:38 GMT time="2022-05-30T14:52:28-04:00" level=info msg="Started successfully, go nuts" component=supervisor-test-respawn
Mon, 30 May 2022 18:52:38 GMT time="2022-05-30T14:52:28-04:00" level=warning msg="Process exited with code: 0" component=supervisor-test-respawn
Mon, 30 May 2022 18:52:38 GMT time="2022-05-30T14:52:28-04:00" level=info msg="respawning in 10ms" component=supervisor-test-respawn
Mon, 30 May 2022 18:52:38 GMT time="2022-05-30T14:52:28-04:00" level=info msg="Restarted (1)" component=supervisor-test-respawn
Mon, 30 May 2022 18:52:38 GMT time="2022-05-30T14:52:28-04:00" level=info msg="Shutting down pid 9782" component=supervisor-test-respawn
Mon, 30 May 2022 18:52:38 GMT --- FAIL: TestRespawn (0.03s)
Mon, 30 May 2022 18:52:38 GMT     supervisor_test.go:155: Respawn failed: supervisor-test-respawn
Mon, 30 May 2022 18:52:38 GMT time="2022-05-30T14:52:28-04:00" level=info msg="Starting to supervise" component=supervisor-test-stop-while-respawn
Mon, 30 May 2022 18:52:38 GMT time="2022-05-30T14:52:28-04:00" level=info msg="Started successfully, go nuts" component=supervisor-test-stop-while-respawn
Mon, 30 May 2022 18:52:38 GMT time="2022-05-30T14:52:28-04:00" level=warning msg="exit status 1" component=supervisor-test-stop-while-respawn
Mon, 30 May 2022 18:52:38 GMT time="2022-05-30T14:52:28-04:00" level=info msg="respawning in 1s" component=supervisor-test-stop-while-respawn
Mon, 30 May 2022 18:52:38 GMT FAIL
Mon, 30 May 2022 18:52:38 GMT FAIL	github.com/k0sproject/k0s/pkg/supervisor	1.072s
```
</details>

--- a/pkg/supervisor/supervisor_test.go
+++ b/pkg/supervisor/supervisor_test.go
@@ -17,8 +17,10 @@ package supervisor
 
 import (
 	"fmt"
+	"io/ioutil"
 	"os"
 	"os/exec"
+	"path/filepath"
 	"sort"
 	"strings"
 	"syscall"
@@ -124,31 +126,48 @@ func TestGetEnv(t *testing.T) {
 }
 
 func TestRespawn(t *testing.T) {
-	truePath, err := exec.LookPath("true")
+	tmpDir := t.TempDir()
+	pingFifoPath := filepath.Join(tmpDir, "pingfifo")
+	pongFifoPath := filepath.Join(tmpDir, "pongfifo")
+
+	err := syscall.Mkfifo(pingFifoPath, 0666)
 	if err != nil {
-		t.Errorf("could not find a path for 'true' executable: %s", err)
+		t.Errorf("Failed to create fifo %s: %v", pingFifoPath, err)
+	}
+	err = syscall.Mkfifo(pongFifoPath, 0666)
+	if err != nil {
+		t.Errorf("Failed to create fifo %s: %v", pongFifoPath, err)
 	}
 
 	s := Supervisor{
 		Name:           "supervisor-test-respawn",
-		BinPath:        truePath,
+		BinPath:        "/bin/sh",
 		RunDir:         ".",
-		Args:           []string{},
-		TimeoutRespawn: 10 * time.Millisecond,
+		Args:           []string{"-c", fmt.Sprintf("cat %s && echo pong > %s", pingFifoPath, pongFifoPath)},
+		TimeoutRespawn: 1 * time.Millisecond,
 	}
 	err = s.Supervise()
 	if err != nil {
 		t.Errorf("Failed to start %s: %v", s.Name, err)
 	}
 
-	// wait til the process exits
-	process := s.GetProcess()
-	for process != nil && process.Signal(syscall.Signal(0)) == nil {
-		time.Sleep(time.Millisecond)
+	// wait til process starts up. fifo will block the write til process reads it
+	err = ioutil.WriteFile(pingFifoPath, []byte("ping 1"), 0644)
+	if err != nil {
+		t.Errorf("Failed to write to fifo %s: %v", pingFifoPath, err)
 	}
 
-	// wait enought time for new process to be respawned
-	time.Sleep((18 * s.TimeoutRespawn) / 10)
+	// save the pid
+	process := s.GetProcess()
+
+	// read the pong to unblock the process so it can exit
+	_, _ = ioutil.ReadFile(pongFifoPath)
+
+	// wait til the respawned process again reads the ping fifo
+	err = ioutil.WriteFile(pingFifoPath, []byte("ping 2"), 0644)
+	if err != nil {
+		t.Errorf("Failed to write to fifo %s: %v", pingFifoPath, err)
+	}
 
 	// test that a new process got re-spawned
 	if process.Pid == s.GetProcess().Pid {

